Feat/contentful demo page

nathanbowang opened the pull request:
Finally, my first PR. 

This PR contains: 
* A new [contentful page](http://localhost:3000/demo/contentful) using `getStaticProps` to load an article from contentful.com at build time. 
* A new getArticleByTitle utility and its test. 
* An `Article` component and its styles and test.
 
Three npm packages were installed:
* `msw` package for mocking a backend server.
* `whatwg-fetch` for using fetch in jest.
* `identity-obj-proxy` for importing CSS module in jest.

Thanks

nathanbowang committed:
feat(contentful-demo-page): add a static hard-coded page

nathanbowang committed:
feat(contentful-demo-page): fetch article at build time

Create a GraphQL query to select the article by title.
Post the query to the Contentful GraphQL Content API.
Get the article at build time by Next's getStaticProps method.
Display the article on the static page.

nathanbowang committed:
feat(contentful-demo-page): move getArticleByTitle to the global utilities
folder

It is not a page. So it should be moved out of ./pages folder. Otherwise `pnpm
build` will fail.

But `pnpm dev` works without any complaint.

nathanbowang committed:
feat(contentful-demo-page): define all inline strings as constants.

Define all inline strings in Contentful component and getArticleByTitle
utility.
Move API/fetch related constants to ./utilities/constants.ts
Adjust whitespaces according to the guide.

nathanbowang committed:
feat(contentful-demo-page): remove .gitkeep

nathanbowang committed:
feat(contentful-demo-page): move getArticleByTitle utility function

Create ./utilities/getArticleByTitle folder.
Move getArticleByTitle utility to the folder.

nathanbowang committed:
feat(contentful-demo-page): add test to getArticleByTitle

Test the `getArticleByTitle()` with an MSW server.
Refactor `GET_ARTICLE_BY_TITLE_QUERY` to
`GET_ARTICLE_BY_TITLE_GRAPHQL_QUERY` to explicit the query language.

nathanbowang committed:
feat(contentful-demo-page): remove `--passWithNoTests` option

nathanbowang committed:
feat(contentful-demo-page): adjust import order and empty lines

nathanbowang committed:
feat(contentful-demo-page): create an Article component

nathanbowang committed:
feat(contentful-demo-page): add some style

nathanbowang committed:
feat(contentful-demo-page): add a test against Article component

Add the test.
Install identity-obj-proxy.
Modify moduleNameMapper object in jest.config.js, so Jest could import css modules.

nathanbowang committed:
feat(contentful-demo-page): add title for the test

chasm commented on components/Article/index.spec.tsx:
```diff
@@ -0,0 +1,8 @@
+import Article from './'
```
Check the style guide. Components should be no higher in the folder hierarchy than they appear. Article will never appear in Clouds, Footer, Header, or Player, but only in Main, so the folder should be `components/Main/Article`.

Really, this is why I wasn't so hot on just copying code over from the other app, which was hacked up in an emergency. Player should be inside Main -- there should only be Header, Main, Footer at the top level based on semantic HTML -- and the Clouds shouldn't really be a component at all. I was in a big hurry. For this site, we are not in a big hurry and should use it to do things _right_. 100% right, no excuses. We can discuss what "right" is, but whatever decision we reach, we should stick to it like glue.

I plan to get rid of Clouds and move Player (and thus EmbeddedContent) into Main as soon as I have time.

nathanbowang replied:
Ok. `Article` should be under Main.

chasm commented on components/Article/index.tsx:
```diff
@@ -0,0 +1,15 @@
+
+export default function Article({ title, content }: ArticleProps): JSX.Element {
+  return (
+    <main className={style.box}>
```
Why `<main>`? There is only one `<main>` element on a page: [A document mustn't have more than one `<main>` element that doesn't have the hidden attribute specified](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/main).

This is an article, so it should use [the `<article>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/article). The `<article>` element is [sectioning content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#sectioning_content).

Edit: Having gotten further, I think we should use the `<article>` element, but this component should be called `PageContent` with `PageContentProps` and `PageContent.module.css` (and moved to the `Main` folder). See below.

nathanbowang replied:
Ok. I see. 
I thought this branch would be discarded so haven't thought much about its relationship with other components and page. I am happy to merge them to a demo branch and leave it for the future.

chasm replied:
Well, that's my point -- if we merge to main, then it _won't_ be discarded. But I also want to be able to deploy it so we can check it out and show it to the client (Jake). So we should set up a second S3 bucket, etc. to deploy a demo app (really a test app, but demo sounds better) as well as the production version. @reubenberghan ?

chasm commented on components/Article/Article.module.css:
```diff
@@ -0,0 +1,14 @@
+.box {
```
Perhaps this should be `article` instead of `box`?

nathanbowang replied:
Yeap.

chasm replied:
Keeping the names in sync makes for less cognitive load, I think.

chasm commented on jest.config.js:
```diff
@@ -4,5 +4,6 @@ module.exports = {
     '^@components(.*)$': '<rootDir>/components$1',
     '^@pages(.*)$': '<rootDir>/pages$1',
     '^@hooks(.*)$': '<rootDir>/hooks$1',
+    '\\.(css)$': 'identity-obj-proxy',
```
What is this?

nathanbowang replied:
Without this module, Jest can's import CSS modules. That says Jest will throw an error at the following line: 
`import style from './Article.module.css`

chasm replied:
Ah! I learned something new. Thanks!

chasm commented on package.json:
```diff
@@ -8,7 +8,7 @@
     "lint": "eslint . --ignore-path .gitignore --ext .js,.ts,.tsx --fix",
     "start": "next start",
-    "test:local": "jest --passWithNoTests",
+    "test:local": "jest",
```
Thanks.

chasm commented on package.json:
```diff
@@ -35,17 +35,20 @@
     "react-icons": "^4.2.0",
-    "tslib": "^2.1.0"
+    "tslib": "^2.1.0",
+    "whatwg-fetch": "^3.6.2"
```
Do we need this? Is this a Dodds recommendation?

nathanbowang replied:
Yes, Node needs a fetch polyfill. Dodds used this one with MSW in the video.

chasm replied:
Still, eh? Well, whatwg is probably the best of them. I figured it was Dodds.

nathanbowang replied:
Yeah, I was also curious about it. And got an error without it. : )

chasm commented on pages/demo/contentful.tsx:
```diff
@@ -0,0 +1,18 @@
+import Article, { ArticleProps } from '../../components/Article'
```
To be clearer, it might be better to do this like this:

```typescript
import Article from '../../components/Main/Article'
import type { ArticleProps } from '../../components/Main/Article'
```

That makes it very clear that one is a component and the other a type. @reubenberghan thoughts?

Edit: Having gotten further and given this more thought, I think you should change this to this:

```typescript
import Page from '../../components/Main/PageContent'
import type { PageContentProps } from '../../components/Main/PageContent'
```

reubenberghan replied:
Yeah I agree with keeping the type imports separate

nathanbowang replied:
OK

chasm commented on pages/demo/contentful.tsx:
```diff
@@ -0,0 +1,18 @@
+import { GetStaticProps } from 'next'
+import React from 'react'
+
+export default function Contentful({ title, content }: ArticleProps): JSX.Element {
```
Check the style guide. _There are no files named anything other than `index.tsx`_. The name of the page goes on the _folder_. And do we want a `demo` folder in production? I am thinking _no_. So maybe we should create a demo branch that we can merge this PR to instead of main? (**Done!**) Then we can dispense with the `demo` folder as the whole thing is a demo:

```sh
pages/contentful/index.tsx
```

This would then be at `the-jac.co.nz/contentful/`. But does that make sense? Aren't we revealing an _implementation_ detail in our URL? What if we switch to some other service?

My guess is that this is intended to be a generic "article" page. In that case, we will want the name of the article to be in the URL. Next.js is all set up to do this. The question is, do we want to put articles at an `articles` URL? Will this site have articles (@jakebax)? Or will we just have pages that will be updated by Jake? I think for now we should assume the latter.

So the actual folder/file structure here should use a [dynamic route](https://nextjs.org/learn/basics/dynamic-routes/page-path-external-data), right?

```sh
pages/
  [slug]/
    index.tsx
  index.tsx
```

This means that the home page is fixed and always present (though it can still load data from contentful), but every other site page is dynamic. We will have to figure out what this means for the contact page, etc. But I'm thinking use MDX and we can inject our components even into markdown pages.

The function should, accordingly, be `Page` or `SitePage` and the `getStaticProps` should retrieve the `slug` name from the path `params`. And as these are pages, not "articles", perhaps the component, while still using the `<article>` element, should be called `PageContent` and the props type, `PageContentProps`. Thoughts?

```typescript
export default function SitePage({ content, slug, title }: PageContentProps): JSX.Element {
  return <PageContent content={content} slug={slug} title={title} />
}

export const getStaticProps: GetStaticProps = async ({ params: { slug } = {}) => {
  const props: PageContentProps = await getPageContentBySlug(slug)

  return { props }
}
```

Titles can be duplicated so we need something unique. Every page should have a `slug` in `train-case` (because it will form part of the URL). You'll need to update Contentful to include a `slug` key-value pair.

(If it's not too onerous, I prefer to alphabetize all lists. -- that includes imports, props, attributes, etc. -- unless there is another order required. Note that I have done so above.)

nathanbowang replied:
OK

nathanbowang replied:
So there should an error handler if the slug is not present in Contentful, right?

chasm replied:
If the slug is not there, then the page is not there, hence it should return null, undefined, [], or whatever (not an error, just no response). Then the SitePage should note that no data was returned and return a 404. But we can set up a page (eventually) to replace the 404 page that apologizes for not finding the page, lists the site pages and/or makes suggestions based on close matches to the slug, or even allows a search... but that's down the road. For now, if the data is not found, have the SitePage output "404 Not Found".

reubenberghan replied:
I know next provides a "Next" 404 out of the box but may want to do something simple for new jac city?

chasm replied:
Sure.

nathanbowang replied:
OK

chasm commented on utilities/constants.ts:
```diff
@@ -0,0 +1,4 @@
+export const APPLICATION_JSON = 'application/json'
```
`APPLICATION_JSON` might be overkill. This isn't "magic" -- it reads exactly what it is. For example, is 3.14 just 3.14 or is it meant to be `PI`? Is 7 just a number or is it `MAXIMUM_NUMBER_OF_ITEMS`? Is `#3cb371` `MEDIUM_SEA_GREEN` or `MEDIUM_AQUA_MARINE`? Without the constant, you don't know.

Another use is to avoid typos. But both `application/json` and `POST` probably don't need constants. @reubenberghan , thoughts?

reubenberghan replied:
I'd agree, these probably don't need to be constants. We can add types for anything that uses these strings which also means we can avoid typos for example:
```
type HttpRequestMethods = 'DELETE' | 'GET' | 'POST' | 'PUT';
```

nathanbowang replied:
OK

chasm replied:
Actually, this is a good candidate for an [Enum](https://www.typescriptlang.org/docs/handbook/enums.html), which TypeScript provides:

```typescript
export enum HttpRequestMethod {
  Delete = 'DELETE',
  Get = 'GET',
  Post = 'POST',
  Put = 'PUT',
}
```

Then you can use the Enum as both the type and to assign values:

```typescript
const method: HttpRequestMethod = HttpRequestMethod.Post
```

Very cool.

nathanbowang replied:
Tried this two way today. They are both enlightening that I haven't thought. In my humble opinion:
* literal type is easier for parameters;
* enum is easier for arguments. 

Please correct me if I missed something.

nathanbowang replied:
```ts
test('declare a litral type of an argument', () => {
  // method is defined as a string in the real fetch()
  type Option = { method?: string }
  function mockFetch(options: Option): void {
    console.log(options.method)
  }
  type HttpRequestMethods = 'GET' | 'POST'

  const method: HttpRequestMethods = 'GUESS' // Type Error
  mockFetch({ method })
  // Type assertion doesn't work, since HttpRequestMethods overlapped with string
  mockFetch({ method: 'GUESS' as HttpRequestMethods }) // No Error
  mockFetch({ method: <HttpRequestMethods>'GUESS' }) // No Error
})

test('declare a enum type of an argument', () => {
  // method is defined as a string in the real fetch()
  type Option = { method?: string }
  function mockFetch(options: Option): void {
    console.log(options.method)
  }

  enum HttpRequestMethod {
    GET = 'GET',
    POST = 'POST',
  }
  mockFetch({ method: HttpRequestMethod.GET })
  mockFetch({ method: 'GET' })
  mockFetch({ method: 'GUESS' }) // No Error
})

test('declare a litral type for a parameter', () => {
  type HttpRequestMethods = 'GET' | 'POST'
  type Option = { method: HttpRequestMethods }
  function myFetch(options: Option): void {
    console.log(options.method)
  }

  myFetch({ method: 'GUESS' }) // Type Error
})

test('declare a enum type for a parameter', () => {
  enum HttpRequestMethod {
    GET = 'GET',
    POST = 'POST',
  }
  type Option = { method: HttpRequestMethod }
  function myFetch(options: Option): void {
    console.log(options.method)
  }

  myFetch({ method: HttpRequestMethod.GET })
  myFetch({ method: 'GET' }) // Type Error
  myFetch({ method: 'GUESS' }) // Type Error
})
```

chasm replied:
Let's group types at the top of the test. Blank line before consts. Then blank line before expects:

(I know this is just to show options, but let's get in the habit of grouping like items and using single blank lines to separate groups.)

```typescript
test('declare a litral type of an argument', () => {
  type Option = { method?: string }
  type HttpRequestMethods = 'GET' | 'POST'

  function mockFetch(options: Option): void {
    console.log(options.method)
  }

  const method: HttpRequestMethods = 'GUESS' // Type Error

  mockFetch({ method })

  // Type assertion doesn't work, since HttpRequestMethods overlapped with string
  mockFetch({ method: 'GUESS' as HttpRequestMethods }) // No Error
  mockFetch({ method: <HttpRequestMethods>'GUESS' }) // No Error
})
```

chasm commented on utilities/getArticleByTitle/index.spec.ts:
```diff
@@ -0,0 +1,35 @@
+afterAll(() => server.close())
+afterEach(() => server.resetHandlers())
+
+describe('a get article test', () => {
```
I'm pretty sure that Dodds has given up on the `describe` blocks and I'm happy to just go with `test`. It used to be that the `describe` blocks printed out nicely (maybe that was mocha?), but for years I haven't seen any benefit to them and have just done them by habit. Maybe it's time to break the habit. So this might be:

```typescript
test('get page content by slug', () => { ... }
```

nathanbowang replied:
OK

chasm commented on utilities/getArticleByTitle/index.ts:
```diff
@@ -0,0 +1,29 @@
+        }
+      }
+    }
+  `
```
Should the collection be called `pages`, `pageContent`, or something else? What do you think? Remember to add `slug` and make it unique.

nathanbowang replied:
I think the response will have only one page without a limitation. So pageContent sounds good to me.

chasm reviewed (changes requested):
This looks pretty good. Mostly minor style guide stuff, but some important naming issues. Let's see what Reuben thinks.

Also, can you change this PR to merge to `demo` rather than `main`. We should set up a `demo` site separately from the production site where we can test out ideas before we implement them in production.

reubenberghan commented on components/Article/index.tsx:
```diff
@@ -0,0 +1,15 @@
+export default function Article({ title, content }: ArticleProps): JSX.Element {
+  return (
+    <main className={style.box}>
+      <h1 className={style.title}>{title}</h1>
```
When updating to the `<article>` element we may also want to consider whether an `<h1>` is the most appropriate heading element for this component. Could be that an `<h2>` would be the best especially if there is a chance that there is going to be more that one `Article` in the page.

chasm replied:
This is a good point. People always misuse headers. There should only be one `<h1>` per page. If there are multiple articles on a page, they should each begin with an `<h2>`. Subheadings under that should be `<h3>`. It is doubtful that we'll ever need to go below `<h3>`.

Too many devs use `hn` tags to control font size. That is an atrocity. The _n_ is used to indicate position in the _hierarchy_. Headings should be properly nested. This is not only good semantics (for machine learning, e.g.) but also good accessibility (to persons with varying abilities).

The question is, in an application with _user-generated_ content, how can one restrict the use of `hn` elements so that they are used correctly? Frankly, I have always thought that they were a mistake. There should be a single `<heading>` tag (the page heading should be called the `<header>`), and `<section>` tags. The page would be divided into sections and subsections with `<section>` tags and each would permit a `<header>` tag, which would automatically set the level of heading, subheading, etc. according to the nesting of the sections.

Actually, the `<header>` element is _not_ limited to one per page at the top of the page (ditto, `<footer>`). Each section could have a `<header>` tag and when HTML5 was first introduced there was some talk of _all_ the `hn` tags being `<h1>`, but nested in proper headers in sections (or articles or asides, which are the same thing -- sectioning elements).

We'll have to think about this. Maybe we can be the first CMS builders to get it right.

reubenberghan commented on utilities/getArticleByTitle/index.ts:
```diff
@@ -0,0 +1,29 @@
+  const res = await fetch(CONTENTFUL_API_URL, {
+    method: POST,
+    headers: {
+      'Content-Type': APPLICATION_JSON,
```
Yeah if we get correct types for `fetch` then you shouldn't need to worry about typos / using constants here

chasm replied:
But the constant for the URL is good, right?

reubenberghan replied:
Yeah the URL constant is still good 😁

reubenberghan commented on utilities/getArticleByTitle/index.ts:
```diff
@@ -0,0 +1,29 @@
+  })
+  const { data } = await res.json()
+  // articleCollection can only contains zero or one article. Since title is an unique field.
+  const article: ArticleProps = data.articleCollection.items[0]
```
Wonder if there's a query where you can get one or none results for a unique key?

chasm replied:
One or none is always possible -- just misspell the key. So we need to allow for that.

chasm replied:
Also, this is a better way of doing the above, I think:

```typescript
const [pageContent]: PageContentProps[] = data.pageContent.items
```

chasm replied:
@natebowang does Contentful return an items array even if you're searching for a single item by a unique key?

nathanbowang replied:
I will work on that. I think Contentful should have a way to just query a record. I am new to GraphQL and will have some research on that. 
Is generate GraphQL query by a query builder better than just strings? Like [this one](https://www.npmjs.com/package/gql-query-builder).

chasm replied:
@reubenberghan may have his own opinion, but I think not. Write GraphQL queries in GraphQL. Write CSS in CSS. Write HTML, to the extent possible, in HTML (or JSX, which is mostly equivalent). Devs who have a hammer and want everything to be a nail frustrate the hell out of me. USE THE RIGHT TOOL FOR THE JOB, FOOLS. (Did you see my rant about CSS-in-JS? Same applies.)

Here is what it looks like with `graphql-tag`:

```typescript
import gql from 'graphql-tag'

export const teamsByMemberIdQuery = gql`
  query TeamsByMemberId($memberId: ID!) {
    teamsByMemberId(memberId: $memberId) {
      displayName
      id
      createdAt
      updatedAt
    }
  }
`
```

Definitely, learning GraphQL well should be part of this exercise.

nathanbowang committed:
feat(contentful-demo-page): fix several PR issues

Fix all pull request #5 issues, except:
* GraphQL query to get one or none
* HttpRequestMethod enum
* dynamic path and slug parameter

nathanbowang committed:
feat(contentful-demo-page): update GraphQL query

* Query page content by its id

nathanbowang committed:
feat(contentful-demo-page): use HttpRequestMethod enum

Substitute constant "POST" by HttpRequestMethod enum.

chasm commented on components/Main/PageContent/index.spec.tsx:
```diff
@@ -0,0 +1,8 @@
+
+test('should show title and content', () => {
+  const { getByText } = render(<PageContent title="JS" content="JavaScript is a programming language." />)
+  expect(getByText('JS'))
```
Can we separate groups of similar code with blank lines? So a single blank line between the variable declarations and the `expect` statements.

chasm commented on utilities/constants.ts:
```diff
@@ -0,0 +1,8 @@
+export const CONTENTFUL_API_ACCESS_TOKEN = 'Bearer qVGC4H5ZmObAQMKyniPJNtNjtGkfmhi_OgVVy9jmtR8'
+export const CONTENTFUL_API_URL = 'https://graphql.contentful.com/content/v1/spaces/zqvr29h6rh51/'
+export enum HttpRequestMethod {
```
Can we insert a blank line here to make clear that one group is `const` and one is `enum`?

nathanbowang committed:
feat(contentful-demo-page): add blank lines

nathanbowang committed:
chore(upgrade-dependencies): update node to 15.12.0

nathanbowang committed:
chore(upgrade-dependencies): update devDependencies and dependencies

nathanbowang committed:
feat(contentful-demo-page): dynamic routes with slug

Use optional-catch-all-routes routes to any arbitary depth except
index page.
Known issue: can't route to index page.

nathanbowang committed:
feat(contentful-demo-page): dynamic routes works with the index page

Fix the issue in the last commit.

reubenberghan commented on pages/[[...slug]]/index.tsx:
```diff
@@ -0,0 +1,43 @@
+  let slugString: string
+
+  if (params?.slug) {
+    // params.slug = [ 'demo', 'js' ]
```
Small thing but just before this is ready to merge could we remove the commented code from this file (I'm assuming it is no longer needed) 😁

natebowang replied:
Hi, that is an example of the data type. Could that be helpful? Maybe I should put it somewhere else?

nathanbowang replied:
Hi, that was an example of the data. I will use types instead of comments.

chasm replied:
This response worries me. Types are not alternatives to comments. Types are required. They are a form of static testing. They ensure that your code is correct -- that you are passing and returning the correct data types. The TypeScript type checker should be running on them and it should not be possible to make a commit or merge to main if the types don't check.

Comments do none of the above. Well-written code should have zero comments. The only reason for a comment is to explain a workaround if it cannot be avoided -- and then only to prevent the next coder (who might be you) from undoing the workaround and then discovering why it was there. As soon as the workaround can be safely removed, it should be and so, too, should the comment.

So no comments. Everything should be typed, and nothing should have type "any".

But there is another problem here and I'm surprised that Reuben missed it. `let` is for mutability. Everything in this code should be immutable, so we should never be using `let` (and absolutely never `var`). Everything should be `const`, but that's not enough. `const` is not immutable, only un-reassignable. So we must be careful not to mutate objects or arrays assigned with `const`.

Also, do not name variables after the type (e.g., `slugString`). Name them what they are, semantically: `path` or `pathname`.

Instead of this:

```typescript
export const getStaticProps: GetStaticProps = async ({ params }) => {
  let slugString: string

  if (params?.slug) {
    slugString = (params?.slug as string[]).join('/')
  } else slugString = 'index'

  const pageContent: PageContentProps = await getPageContentBySlugString(slugString)

  return { props: { pageContent } }
}
```

This:

```typescript
export const getStaticProps: GetStaticProps = async ({ params: { slug } } = {}) => {
  const path = slug ? slug.join('/') : 'index'

  const pageContent: PageContentProps = await getPageContentByPath(path)

  return { props: { pageContent } }
}
```

And this:

```typescript
export const getStaticPaths: GetStaticPaths = async () => {
  const allPageContentsWithSlugString: PageContentsWithSlugString[] =
    await getAllPageContentsWithSlugString()
  let paths = allPageContentsWithSlugString.map(({ slugString }) => ({ params: { slug: slugString.split('/') } }))

  paths = paths.concat({ params: { slug: [] } })

  return {
    fallback: false,
    paths,
  }
}
```

Should be, um, wait. What is this doing? Why do we need this?

If we need this, then the way to do it would be this:

```typescript
const slugs = [
   ...allPageContents.map(({ path }) => ({ params: { slug: path.split('/') } })),
  { params: { slug: [] } }
]
```

No reassignment, please.

You could also have done this:

```typescript
  const slugs = allPageContents.map(({ path }) => ({ params: { slug: path.split('/') } }))

  return {
    fallback: false,
    paths: slugs.concat({ params: { slug: [] } }) // is paths the right name?
  }
```

Or even this:

```typescript
export const getStaticPaths: GetStaticPaths = async () => {
  const allPageContents: PageContents[] = await getAllPageContents()

  return {
    fallback: false,
    paths: [
      ...allPageContents.map(({ path }) => ({ params: { slug: path.split('/') } })),
      { params: { slug: [] } }
    ]
  }
}
```

But what is the purpose of `getStaticPaths`? Is this a Next.js thing or a Contentful thing or something custom?

reubenberghan replied:
Oh yeah I did miss the use of `let` and reassignment but yeah that code should be changed :grin:

`getStaticPaths` is a Next.js thing. For a static site this function generates a page for each path in the returned array.

chasm replied:
So I see. But what is Next.js using it for? Why does it need it? Is it for static site generation?

I'm guessing that's the purpose (I don't remember using it with SSR before). Makes sense that it would need to know what pages to generate. Hmmm.

It would be interesting to know whether Contentful can a) store actual string arrays, and b) use them as a primary (or secondary) key, so that we can use the slug array for look up. That would avoid casting back and forth. And if we have to return it as an array anyway, maybe `stringify` and `parse` are a better way to go.

chasm replied:
BTW, does this work for all pages (i.e., we don't need the `pages/index.tsx` file anymore?) That would be cool.

nathanbowang replied:
It works for all pages. We don't need the pages/index.tsx anymore.

chasm replied:
Nice.

nathanbowang replied:
[getStaticPaths](https://nextjs.org/docs/basic-features/data-fetching#getstaticpaths-static-generation) is required for dynamic routes:
> If a page has dynamic routes (documentation) and uses getStaticProps it needs to define a list of paths that have to be rendered to HTML at build time.

So I think that means if we use `params.slug` in the getStaticProps, then it is mandatory to have this function.

reubenberghan commented:
I can't seem to add a comment to the [`pages/index.tsx` filename change](https://github.com/the-bax/new-jac-city/pull/5/files#diff-f2d6f7ba182cf5703b4580fcae3509561aa66e3b99242b3272e75a54a2f190bd) as it's just a rename but wondering whether we want to leave this here? I would think it could just be deleted as part of this PR unless we want to leave it as the index route?

chasm reviewed (changes requested):
Please remove all `let` variables and check to make sure everything is immutable. Careful with names. PageContent should _include_ the path, so there is no need to put `path` in the query or function name, unless it is your primary key (e.g., `getPageContentByPath`).

nathanbowang committed:
feat(contentful-demo-page): reorganize api utilities

Amended as Reuben said in the PR
Write a generic fetchGraphQl function.
Combine getAllPageContentsWithSlugString and getPageContentBySlugString
in one file useing the fetchGraphQl function.
Combine the corresponding test.

nathanbowang committed:
feat(contentful-demo-page): continue to reorganize api utilities

Give up the separate `variables` of the GraphQL query. That introduced too
much complexity like condidional type. The query string will not be
manipulate by users, so it is safe to remove. On top of that, remove
it makes the query more clear to read.
Add some nullish check.

nathanbowang committed:
feat(contentful-demo-page): polish some types

`getPageContentBySlugString()` should return undefined if the pageContent
record can not be found.
`getAllPageContentsWithSlugString()` should return [] if no pageContents
record exist.

natebowang commented:
Here we go. That's all I want to put in this PR and I amended it as you suggested. This PR is a little bit too big now.

chasm commented:
Hi, Nate,

I'm still seeing `let` in this file: https://github.com/the-bax/new-jac-city/blob/dfc2321b37253e3571bdd713b0522f46887f70fd/pages/%5B%5B...slug%5D%5D/index.tsx

Am I missing an update (GitHub is annoyingly difficult to navigate lately)? There is a long comment from me and a short concurrence from Reuben above that needs to be fixed. No reassigned variables and no mutability, please. Also, some naming issues. Check the above comment out and let me know if I'm missing something.

natebowang commented:
Oh, I missed your last comment. But I already deleted the long comment.

chasm commented:
Confused by `deleted the long comment`. Are you talking comments here in the PR or comments in the code? Please don't delete any comments on the PR.

We're almost there!

nathanbowang commented on pages/[[...slug]]/index.tsx:
```diff
@@ -29,10 +28,12 @@ export const getStaticProps: GetStaticProps = async ({ params }) => {
+
   const allPageContentsWithSlugString: PageContentsWithSlugString[] = await getAllPageContentsWithSlugString()
-  // paths: [{ params: { slug: ['demo', 'js'] } }, { params: { slug: ['demo', 'python'] } }],
-  let paths = allPageContentsWithSlugString.map(({ slugString }) => ({ params: { slug: slugString.split('/') } }))
```
I deleted the comments in the code. 
I see what's the problem. Probably because there are too many commits in this PR which makes it very hard to track the changes in the 'File changed' tab. Sorry for making this inconvenience. 
Anyway, I will push further commits to deal with the unsolved problems.

nathanbowang committed:
feat(contentful-demo-page): substitute let by const

natebowang commented:
I will push more commits to solve all problems. Please wait for a while.

nathanbowang committed:
feat(contentful-demo-page): rename slugString to pathname

Rename slugString and API utility function names to increase readability.
Continue check types.

nathanbowang committed:
feat(contentful-demo-page): fix a bug in getStaticPaths

`{params: {slug: []}}` for the path '/' should be generate from pathname
'index', instead of concat to the result. The previous concat way
actually generate two paths, '/' and '/index'.

nathanbowang committed:
Merge branch 'chore/upgrade-dependencies' into feat/contentful-demo-page

natebowang commented on utilities/api/index.spec.ts:
```diff
@@ -1,39 +1,39 @@
       case /{\s*content\s*title\s*}/.test(query):
-        items = [pageContent]
+        items = [mockPageContent]
+        break
```
Still have one `let` in the test. This part should be rewritten using mocking GraphQL API from MSW. So this `let` will be removed then. 
But still curious, how to avoid using let in this scenario：
```js
let x
switch (someString) {
  case 'apple':
    x = 1
    break
  case 'banana':
    x = 0
    break
  case 'orange':
    x = 0
    break
}
```

nathanbowang commented on utilities/api/index.spec.ts:
```diff
@@ -1,39 +1,39 @@
-    const query = (req?.body as Record<string, string>)?.query
-    let items
+    const query = (req.body as Record<string, string>).query
+    let items: PageContentProps[] | Pathname[]
```
Still have one `let` in the test. This part should be rewritten using mocking GraphQL API from MSW. So this `let` will be removed then. 
But still curious, how to avoid using let in this scenario：
```js
let x
switch (someString) {
  case 'apple':
    x = 1
    break
  case 'banana':
    x = 0
    break
  case 'orange':
    x = 0
    break
}
```

chasm replied:
Move the switch into a function:

```typescript
function setXByFruit(fruit) {
  switch (fruit) {
    case 'apple':
      return 1
    case 'banana':
    case 'orange':
      return 0
  }
}
```

Then wherever you need to set x:

```typescript
  const x = setXByFruit(fruit)
```

chasm replied:
I'm looking at your actual code now and this is a very bad use of `switch`. If your switch is on `true`, then aren't you just testing conditions? The proper syntax for that is if/else. You can still move it into a function (you'll need to).

A general rule: if you're doing something very clever, then you're doing something very wrong. Good code is not clever, it's obvious and easy to understand. Despite having programmed for forty years in a couple dozen different languages, I had to think about the code for a moment to figure out what the hell it was doing. That's not a code smell. It's a code _stink_.

I know clever code looks fun, but trust me: [Murder _all_ your darlings](https://tinhouse.com/murder-your-darlings/).

natebowang commented:
That's it. I think I amended all.

reubenberghan commented on pages/[[...slug]]/index.tsx:
```diff
@@ -0,0 +1,35 @@
+export const getStaticPaths: GetStaticPaths = async () => {
+  type ParamWithSlug = { params: { slug: string[] } }
+
+  const allPathnames = (await getAllPathnames()) as Pathname[]
```
I'm interested as to why we need to cast the return value of our own utility function here. Shouldn't we just be returning the type we expect from our own function?

chasm replied:
`as` is generally a code smell and should be avoided as much as possible.

nathanbowang replied:
`getAllPathname()` is an API utility function, and it is possible that no record in the Contentful backend, so it returns `Promise<Pathname[] | []>`. 
`getStaticPaths()` is a function for generating dynamic paths. Since we put all paths in `[[...slug]]`, if `allPathnames` is `[]`, the website won't have any path, which doesn't make sense for a website. So I assert the type could only be `Pathname[]`. 
That is just how I considered it. Is that right?

nathanbowang replied:
I learned from Chas's new code and realized `getAllPathname()` should returns `Promise<Pathname[]>` not `Promise<Pathname[] | []>`. This is not a valid discussion now. Thanks for the advice.

reubenberghan commented on pages/[[...slug]]/index.tsx:
```diff
@@ -0,0 +1,35 @@
+}
+
+export const getStaticPaths: GetStaticPaths = async () => {
+  type ParamWithSlug = { params: { slug: string[] } }
```
Never seen this before. Are we creating a type in the body of the function here?

Could this be at the top of the file with the other types?

nathanbowang replied:
This type was `type Path = { params: { slug: string[] } }` before. Its a Next thing just for the `getStaticPaths()`, so it was put inside of the function. 
It was renamed because I thought `Path` should mean a string, not an object. I can't change the `paths` in the returned object, so I renamed the type. 
I come up with another way to get rid of this type. May be clearer:
```ts
export const getStaticPaths: GetStaticPaths = async () => {
  // type ParamWithSlug = { params: { slug: string[] } }

  const allPathnames = (await getAllPathnames()) as Pathname[]
  const allSlugs: { slug: string[] }[] = allPathnames.map(({ pathname }) => ({
    slug: pathname !== 'index' ? pathname.split('/') : [],
  }))

  // const paths: ParamWithSlug[] = allPathnames.map(({ pathname }) => ({
  //   params: { slug: pathname !== 'index' ? pathname.split('/') : [] },
  // }))

  return {
    fallback: false,
    paths: allSlugs.map((slug) => ({ params: slug })),
  }
}
```
How do you feel about it?

nathanbowang replied:
This is not a valid discussion now. Thanks for the advice.

reubenberghan commented on pages/[[...slug]]/index.tsx:
```diff
@@ -0,0 +1,35 @@
+}
+
+export const getStaticProps: GetStaticProps = async ({ params }) => {
+  const pathname: string = (params?.slug as string[])?.join('/') ?? 'index'
```
Annoying that we have to cast the `params?.slug` here.

Could we do something like?
```
const pathname = Array.isArray(params?.slug) ? params.slug.join('/') : 'index';
```

nathanbowang replied:
I tried it and that works. However, TS threw a warning. It seems TS doesn't know we already check that `params` is defined and `slug` is an array. 
![image](https://user-images.githubusercontent.com/30286194/112219511-f7c19800-8c89-11eb-970d-4041d4676f5d.png)
So I have to add another casting to promise TS `params` must be a `ParsedUrlQuery` here. But TS keeps on complaining `slug` could be a string, which is so dumb. And then I have to promise `slug` has to be a string array. The code ended up like this:
![image](https://user-images.githubusercontent.com/30286194/112222459-6fdd8d00-8c8d-11eb-8def-821e1c24b2b0.png)

I read about the difference between type assertation and casting from [here](https://basarat.gitbook.io/typescript/type-system/type-assertion#type-assertion-vs-casting). The author deems type assertation as a way to promise TS the type has been checked. I don't know the other ways to solve these warnings and willing to know more.

nathanbowang replied:
This is not a valid discussion now. Thanks for the advice.

chasm commented on pages/[[...slug]]/index.tsx:
```diff
@@ -0,0 +1,35 @@
+  const allPathnames = (await getAllPathnames()) as Pathname[]
+  const paths: ParamWithSlug[] = allPathnames.map(({ pathname }) => ({
+    params: { slug: pathname !== 'index' ? pathname.split('/') : [] },
+  }))
```
We can do better than this, but for the record we should never use this:

```typescript
{ slug: pathname !== 'index' ? pathname.split('/') : [] }
```

But instead:

```typescript
{ slug: pathname === 'index' ? [] : pathname.split('/') }
```

Never choose the negative when the positive will do.

chasm commented on pages/[[...slug]]/index.tsx:
```diff
@@ -0,0 +1,35 @@
+    fallback: false,
+    paths,
+  }
+}
```
So I spent some time trying to come up with the cleanest way to do this. We can discuss it on Slack when we get a chance. But here's what I came up with:

```typescript
import * as React from 'react'
import PageContent from '../../components/Main/PageContent'
import { getAllPathnames, getPageContentByPathname } from '../../utilities/api'
import type { PageContentProps } from '../../components/Main/PageContent'
import type { Pathname } from '../../utilities/api'

export type SitePageProps = {
  pageContent: PageContentProps
}

export type PageParams = {
  params: {
    slug: string[]
  }
}

export type StaticProps = {
  props: {
    pageContent?: Record<string, string>
  }
}

export type PagePath = {
  params: {
    slug: string[]
  }
}

export type StaticPaths = {
  fallback: boolean
  paths: PagePath[]
}

export default function SitePage({ pageContent }: SitePageProps): JSX.Element {
  return <PageContent content={pageContent.content} title={pageContent.title} />
}

export const getStaticProps = async ({ params: { slug } }: PageParams): Promise<StaticProps> => {
  const pathname: string = slug?.join('/') ?? 'index'
  const pageContent = await getPageContentByPathname(pathname)

  return { props: { pageContent } }
}

export const getStaticPaths = async (): Promise<StaticPaths> => {
  const pathnames: Pathname[] = await getAllPathnames()

  return {
    fallback: false,
    paths: pathnames.map<PagePath>(({ pathname }) => ({
      params: {
        slug: pathname === 'index' ? [] : pathname.split('/'),
      },
    })),
  }
}
```

Thoughts? Gripes? Insults?

chasm commented on utilities/api/index.ts:
```diff
@@ -0,0 +1,60 @@
+type CollectionData = { pageContentCollection: { items: Pathname[] } }
+type ItemData = { pageContentCollection: { items: [PageContentProps] } }
+type Data = CollectionData | ItemData
+export type Pathname = { pathname: string }
```
Let's export all types so that they can easily be imported by anyone who needs them. Also, let's not depend on hoisting if we don't have to:

```typescript
export type Pathname = { pathname: string }
export type CollectionData = { pageContentCollection: { items: Pathname[] } }
export type ItemData = { pageContentCollection: { items: PageContentProps[] } }
export type Data = CollectionData | ItemData
```

Also, `PageContentProps[]` not `[PageContentProps]`.

Finally, this naming is quite confusing. What is `CollectionData`? What is `ItemData`? These seem very different, so why do we need the union type?

chasm commented on utilities/api/index.ts:
```diff
@@ -0,0 +1,60 @@
+  const { data }: { data: Data } = await res.json()
+
+  return data
+}
```
Do we really need such an abstract function? Doesn't this somewhat defeat the purpose of types? This could return `CollectionData` _or_ `ItemData`, but those are very different things. Shouldn't we be clear what we expect? Then below we find ourselves using the code smell of `as` to force the type back to what we want.

It feels like this is doing too many things.

chasm commented on utilities/api/index.ts:
```diff
@@ -0,0 +1,60 @@
+
+  return data.pageContentCollection.items
+}
+
```
What if we did this instead?

```typescript
export type Path = { pathname: string }

const headers = {
  Authorization: CONTENTFUL_API_ACCESS_TOKEN,
  'Content-Type': 'application/json',
}

export async function listPagePaths(): Promise<Path[]> {
  const body = JSON.stringify({
    operationName: 'ListPagePaths',
    query: `query ListPagePaths {
      pageContentCollection {
        items {
          pathname
        }
      }
    }`,
  })

  const result = await fetch(CONTENTFUL_API_URL, {
    method: HttpRequestMethod.POST,
    headers,
    body,
  })

  const json = await result.json()

  return json?.data?.pageContentCollection?.items || []
}
```

This is a first pass at it, but some thoughts:

- `{ pathname: string }` is an _object_ with a pathname property, not the pathname, so the type should not be `Pathname` as that's very confusing (it confused me repeatedly).
- The headers are always the same, so set them outside the functions
- Generate the body as a JSON string separately to reduce noise. Now we can see what we're doing.
- Now the call to `fetch` is very simple and clear.
- We don't need to split the json out only to end up digging down into `data` anyway.
- Let's return an empty array (or should we throw?) if anything is missing.

I think that we should use `list` when we are returning all objects (or a filtered set) and `get` for single objects. That makes it very clear what we're doing. So `listPagePaths` means list all the page paths.

chasm commented on utilities/api/index.ts:
```diff
@@ -0,0 +1,60 @@
+  const data: ItemData = <ItemData>await fetchGraphQl(operationName, query)
+
+  return data.pageContentCollection.items?.[0]
+}
```
Here is a cleaner approach to this module.

First, the `api` doesn't belong in `utilities`. It's a top level folder so let's move it up.

Next, let's break the functions out into their own folders each with its own test and use the default export. We can move common types into `types.ts` and constants into `constants.ts`

In `/api/constants.ts`:

```typescript
export const CONTENTFUL_API_ACCESS_TOKEN = 'Bearer qVGC4H5ZmObAQMKyniPJNtNjtGkfmhi_OgVVy9jmtR8'
export const CONTENTFUL_API_URL = 'https://graphql.contentful.com/content/v1/spaces/zqvr29h6rh51/'

export enum HttpRequestMethod {
  GET = 'GET',
  POST = 'POST',
  DELETE = 'DELETE',
  PUT = 'PUT',
}

export const headers = {
  Authorization: CONTENTFUL_API_ACCESS_TOKEN,
  'Content-Type': 'application/json',
}
```

In `/api/types.ts`:

```typescript
export type Path = { pathname: string }
```

In `/api/getPageContentByPathname/index.ts`:

```typescript
import { CONTENTFUL_API_URL, headers, HttpRequestMethod } from '../constants'
import type { PageContentProps } from '../../components/Main/PageContent'

export default async function getPageContentByPathname(pathname: string): Promise<PageContentProps | undefined> {
  const body = JSON.stringify({
    operationName: 'GetPageContentByPathname',
    query: `query GetPageContentByPathname {
      pageContentCollection(where: {pathname: "${pathname}"}) {
        items {
          content
          title
        }
      }
    }`,
  })

  const result = await fetch(CONTENTFUL_API_URL, {
    method: HttpRequestMethod.POST,
    headers,
    body,
  })

  const json = await result.json()
  const [pageContent] = json?.data?.pageContentCollection?.items || []

  return pageContent
}
```

In `/api/getPageContentByPathname/index.test.ts`:

```typescript
import getPageContentByPathname from './';
import { CONTENTFUL_API_URL } from '../constants';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import 'whatwg-fetch';
import type { PageContentProps } from '../../components/Main/PageContent'

const mockPageContent: PageContentProps = {
  content: 'mock content',
  title: 'mock title',
}

const server = setupServer(
  rest.post(CONTENTFUL_API_URL, (_, res, ctx) => {
    const items: PageContentProps[] = [mockPageContent]

    return res(
      ctx.json({
        data: {
          pageContentCollection: {
            items,
          },
        },
      }),
    )
  }),
)

test('get page content by pathname', async () => {
  server.listen({ onUnhandledRequest: 'error' })

  expect(await getPageContentByPathname('mockPathname')).toMatchObject(mockPageContent)

  server.close()
})
```

In `/api/listPagePaths/index.ts`:

```typescript
import { CONTENTFUL_API_URL, headers, HttpRequestMethod } from '../constants'
import type { Path } from '../types'

export default async function listPagePaths(): Promise<Path[]> {
  const body = JSON.stringify({
    operationName: 'ListPagePaths',
    query: `query ListPagePaths {
      pageContentCollection {
        items {
          pathname
        }
      }
    }`,
  })

  const result = await fetch(CONTENTFUL_API_URL, {
    method: HttpRequestMethod.POST,
    headers,
    body,
  })

  const json = await result.json()

  return json?.data?.pageContentCollection?.items || []
}
```

In `/api/listPagePaths/index.test.ts`:

```typescript
import listPagePaths from './'
import { CONTENTFUL_API_URL } from '../constants'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import 'whatwg-fetch'
import type { Path } from '../types'

const mockAllPaths: Path[] = [
  {
    pathname: 'demo/js',
  },
  {
    pathname: 'demo/python',
  },
]

const server = setupServer(
  rest.post(CONTENTFUL_API_URL, (_, res, ctx) => {
    const items: Path[] = mockAllPaths

    return res(
      ctx.json({
        data: {
          pageContentCollection: {
            items,
          },
        },
      }),
    )
  }),
)

test('list page paths', async () => {
  server.listen({ onUnhandledRequest: 'error' })

  expect(await listPagePaths()).toMatchObject(mockAllPaths)

  server.close()
})
```

In `/pages/_types.ts`:

```typescript
import type { PageContentProps } from '../components/Main/PageContent'

export type SitePageProps = {
  pageContent: PageContentProps
}

export type PageParams = {
  params: {
    slug: string[]
  }
}

export type StaticProps = {
  props: {
    pageContent?: Record<string, string>
  }
}

export type PagePath = {
  params: {
    slug: string[]
  }
}

export type StaticPaths = {
  fallback: boolean
  paths: PagePath[]
}
```

Finally, I had to update `/pages/[[...slug]]/index.ts` thus:

```typescript
import * as React from 'react'
import getPageContentByPathname from '../../api/getPageContentByPathname'
import listPagePaths from '../../api/listPagePaths'
import PageContent from '../../components/Main/PageContent'
import type { Path } from '../../api/types'
import type { PageParams, PagePath, SitePageProps, StaticPaths, StaticProps } from '../_types'

export const getStaticProps = async ({ params: { slug } }: PageParams): Promise<StaticProps> => {
  const pathname: string = slug?.join('/') ?? 'index'
  const pageContent = await getPageContentByPathname(pathname)

  return { props: { pageContent } }
}

export const getStaticPaths = async (): Promise<StaticPaths> => {
  const pagePaths: Path[] = await listPagePaths()

  return {
    fallback: false,
    paths: pagePaths.map<PagePath>(({ pathname }) => ({
      params: {
        slug: pathname === 'index' ? [] : pathname.split('/'),
      },
    })),
  }
}

export default function SitePage({ pageContent }: SitePageProps): JSX.Element {
  return <PageContent content={pageContent.content} title={pageContent.title} />
}
```

I prefer to put the default export at the bottom.

Page works, no errors, tests pass. What did I miss?

I think that this is much cleaner. The names more clearly state what each thing does. Each thing does just one thing. Files are (mostly) shorter and easier to see at a glance. The folder structure better mirrors the module structure. The tests are simpler and we don't need `before` and `after` (I try to avoid them). Etc.

@reubenberghan did I get anything wrong? @natebowang do you see what I'm getting at?

reubenberghan replied:
All looks pretty good to me, nice and clean. My only note is re the `HttpRequestMethod` enum and potentially moving that into the `types.ts` file and importing accordingly. Unless I'm missing something about that...

chasm reviewed (changes requested):
Well, damn. Found more stuff. I really need to find time to go through this, but this is great for learning. If you have questions about anything, @natebowang , please ask!

chasm committed:
refactor: update api as example

chasm committed:
chore: remove comment

nathanbowang committed:
fix: update getStaticPaths output type to include undefined

nathanbowang committed:
fix: deleted ./pages/_types.ts

Next will throw error if ./pages folder contains any file without a
React Component exported when build. The types have been moved into the
index page file.

chasm committed:
refactor: move page types into types folder

chasm committed:
Merge pull request #8 from the-bax/refactor/redo-demo

refactor: update api as example

reubenberghan commented on api/getPageContentByPathname/index.ts:
```diff
@@ -0,0 +1,27 @@
+import { CONTENTFUL_API_URL, headers, HttpRequestMethod } from '../constants'
```
`HttpRequestMethod` is a type so might be good to explicitly import it this way. Thinking about it maybe this type should be in the `types.ts` rather than the `constants.ts` as well.

reubenberghan commented on api/listPagePaths/index.ts:
```diff
@@ -0,0 +1,25 @@
+import { CONTENTFUL_API_URL, headers, HttpRequestMethod } from '../constants'
```
Same here re the explicit type import for the `HttpRequestMethod` enum 😁

reubenberghan reviewed (changes requested):
Looking pretty damn close, nice work @natebowang!

nathanbowang committed:
refactor: move HttpRequestMethod into types.ts

nathanbowang committed:
refactor: remove HttpRequestMethod from constants.ts

nathanbowang committed:
test(mock-graphql-api): replace mock REST API by mock GraphQL API

reubenberghan commented on api/getPageContentByPathname/index.ts:
```diff
@@ -0,0 +1,28 @@
+import { CONTENTFUL_API_URL, headers } from '../constants'
+import { HttpRequestMethod } from '../types'
```
Ah right that makes sense that we can't import an `enum` explicitly as a type when we use it as a value 😜

chasm replied:
True dat. I was thinking whether it belongs in the types file, but as it's a TS thing, I'm thinking that that's as good a place as any.

reubenberghan reviewed (approved):
Awesome work @natebowang!

nathanbowang committed:
test: make up tests for short-circuit evaluation

Coverage is 100% now.

nathanbowang committed:
refactor: update test/test-utils to ts

nathanbowang committed:
test: create end-to-end test

nathanbowang committed:
refactor: move constants into a separate file

nathanbowang committed:
refactor: rename constant.ts to constants.ts

nathanbowang committed:
test: add firefox

nathanbowang committed:
refactor: replace `it` by `test`

nathanbowang committed:
refactor: replace `it` by `test`

nathanbowang committed:
test: replace the unit tests by an integration test

nathanbowang committed:
refactor: add type of response body and delete the loop over paths

nathanbowang committed:
test: remove `expect` involve implementation details

nathanbowang committed:
Merge pull request #9 from the-bax/test/mock-graphql-api

Test/mock graphql api

nathanbowang committed:
refactor: rename the e2e test

Rename the e2e test file.
Delete unnecessary tests.
Move packages for test to devDependencies.

nathanbowang committed:
fix: remove implementation details from e2e test

nathanbowang committed:
refactor: import ReactElement as type

nathanbowang committed:
Merge pull request #10 from the-bax/test/e2e

Test/e2e
